environment_switcher/lib/src/environment.dart

```diff
@@ -0,0 +1,31 @@
+import 'package:equatable/equatable.dart';
+import 'package:flutter/material.dart';
+
+/*
+ * The text in here doesn't need to be localized. It's just 
+ * for testing builds and won't make it to the proper live release
+ */
+
+class Environment extends Equatable {
+  final String name;
+  final String description;
+  final Color bannerColor;
+  final String databaseName;
```

Comment:
This is quite specific to SP, I'm not sure if it's required for a generic solution. 

Does this provide a good way to add extra fields? Subclassing might be a bit awkward.

Reply:
Good points! I've removed the `databaseName` and `value` string parameters, and replaced with a `Map<String, dynamic>` that allows adding as many additional fields as required. Would this work better?

Reply:
Potentially, although it's not type-safe. And you couldn't assert that something you want to be in the map is actually in there. 

My first thoughts were making `Environment` abstract, and make `EnvironmentSwitcher` generic of the form `EnvironmentSwitcher<T extends BaseEnvironment>`. That way we can have `SPEnvironment extends BaeEnvironment` and add the `databaseName` parameter. 

Or is suppose Maybe it doesn't need to be abstract, as `Environment` will work as-is.

Reply:
Ok. In general, do we want to move away from using `Map<String, dynamic>` due it not being type-safe?

I've created an abstract class called `EnvironmentData`. This class should be extended with the necessary fields and their corresponding assertions. In the `Environment` class, there is now an optional field for `EnvironmentData`. How does that sound?

Reply:
> Ok. In general, do we want to move away from using `Map<String, dynamic>` due it not being type-safe?

Yep, if a proper class can be used, it should. 

> I've created an abstract class called `EnvironmentData`. This class should be extended with the necessary fields and their corresponding assertions. In the `Environment` class, there is now an optional field for `EnvironmentData`. How does that sound?

Sounds good. Does that mean that we'd have too do: `(environment.environmentData as FooEnvironmentdata).bar`?

Reply:
No, you would only have to do `environment.environmentData.bar`. I've added a load of generic types extending `EnvironmentData`, which should allow you to just simply call the data variables without having to typecast.